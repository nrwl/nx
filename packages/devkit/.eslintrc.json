{
  "extends": "../../.eslintrc.json",
  "ignorePatterns": ["!**/*", "node_modules"],
  "overrides": [
    {
      "files": ["*.ts", "*.tsx", "*.js", "*.jsx"],
      "rules": {
        "@typescript-eslint/explicit-module-boundary-types": "error",
        "@typescript-eslint/no-restricted-imports": [
          "error",
          {
            "paths": [
              "@nx/workspace",
              "@angular-devkit/core",
              "@angular-devkit/architect",
              "@angular-devkit/schematics"
            ],
            "patterns": [
              {
                "group": [
                  "nx/bin/*",
                  "nx/src/*",
                  "nx/plugins/*",
                  "!nx/src/devkit-internals",
                  "!nx/src/devkit-exports"
                ],
                "allowTypeImports": true,
                "message": "Only import from nx/src/devkit-internals or nx/src/devkit-exports"
              },
              {
                "group": ["@nx/devkit/**/*"],
                "message": "Use a relative import"
              }
            ]
          }
        ]
      }
    },
    {
      "files": ["*.spec.ts"],
      "rules": {
        "@typescript-eslint/no-restricted-imports": "off"
      }
    },
    {
      "files": ["*.ts", "*.tsx"],
      "rules": {}
    },
    {
      "files": ["*.js", "*.jsx"],
      "rules": {}
    },
    {
      "files": ["./package.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/nx-plugin-checks": "error"
      }
    },
    {
      "files": ["./package.json"],
      "parser": "jsonc-eslint-parser",
      "rules": {
        "@nx/dependency-checks": [
          "error",
          {
            "buildTargets": ["build-base"],
            "ignoredDependencies": [
              "nx",
              "typescript",
              "prettier",
              "tslib",
              // Installed to workspace by plugins
              "rxjs",
              "@angular-devkit/core",
              "@angular-devkit/architect",
              "@angular-devkit/schematics",
              "webpack"
            ]
          }
        ]
      }
    }
  ]
}
