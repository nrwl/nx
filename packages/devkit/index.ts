/**
 * The Nx Devkit is the underlying technology used to customize Nx to support
 * different technologies and custom use-cases. It contains many utility
 * functions for reading and writing files, updating configuration,
 * working with Abstract Syntax Trees(ASTs), and more.
 *
 * As with most things in Nx, the core of Nx Devkit is very simple.
 * It only uses language primitives and immutable objects
 * (the tree being the only exception).
 *
 * @module @nx/devkit
 */

// eslint-disable-next-line @typescript-eslint/no-restricted-imports
export * from 'nx/src/devkit-exports';
export * from './public-api';
