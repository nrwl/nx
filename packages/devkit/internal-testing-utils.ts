/* eslint-disable @nx/enforce-module-boundaries */
export * from '../nx/src/internal-testing-utils/assert-valid-migrations';
export * from '../nx/src/internal-testing-utils/run-migration-against-this-workspace';
export * from '../nx/src/internal-testing-utils/with-environment';
export * from '../nx/src/internal-testing-utils/temp-fs';
/* eslint-enable @nx/enforce-module-boundaries */
export { setCwd } from './src/generators/artifact-name-and-directory-utils';
