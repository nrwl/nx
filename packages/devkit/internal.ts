export { signalToCode } from 'nx/src/devkit-internals';
