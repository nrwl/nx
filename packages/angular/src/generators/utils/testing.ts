import type { Tree } from '@nx/devkit';
import { names, readProjectConfiguration } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Linter } from '@nx/eslint';
import { UnitTestRunner } from '../../utils/test-runners';
import { applicationGenerator } from '../application/application';
import type { Schema as ApplicationOptions } from '../application/schema';
import { componentGenerator } from '../component/component';
import { host } from '../host/host';
import type { Schema as HostOptions } from '../host/schema';
import { libraryGenerator } from '../library/library';
import type { Schema as LibraryOptions } from '../library/schema';
import { remote } from '../remote/remote';
import type { Schema as RemoteOptions } from '../remote/schema';

export async function generateTestApplication(
  tree: Tree,
  options: ApplicationOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await applicationGenerator(tree, {
    ...options,
  });
}

export async function generateTestHostApplication(
  tree: Tree,
  options: HostOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await host(tree, { ...options });
}

export async function generateTestRemoteApplication(
  tree: Tree,
  options: RemoteOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await remote(tree, { ...options });
}

export async function generateTestLibrary(
  tree: Tree,
  options: LibraryOptions
): Promise<void> {
  tree.write('.gitignore', '');
  await libraryGenerator(tree, {
    ...options,
  });
}

export async function createStorybookTestWorkspaceForLib(
  libName: string
): Promise<Tree> {
  let tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  tree.write('.gitignore', '');

  await libraryGenerator(tree, {
    directory: libName,
    buildable: false,
    linter: Linter.EsLint,
    publishable: false,
    simpleName: false,
    skipFormat: true,
    unitTestRunner: UnitTestRunner.Jest,
    standalone: false,
  });

  await componentGenerator(tree, {
    name: 'test-button',
    path: `${libName}/src/lib/test-button/test-button`,
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/test-button/test-button.component.ts`,
    `import { Component, Input } from '@angular/core';

export type ButtonStyle = 'default' | 'primary' | 'accent';

@Component({
  selector: 'proj-test-button',
  templateUrl: './test-button.component.html',
  styleUrls: ['./test-button.component.css']
})
export class TestButtonComponent {
  @Input('buttonType') type = 'button';
  @Input() style: ButtonStyle = 'default';
  @Input() age?: number;
  @Input() isOn = false;
}`
  );

  tree.write(
    `${libName}/src/lib/test-button/test-button.component.html`,
    `<button [attr.type]="type" [ngClass]="style"></button>`
  );

  const modulePath = `${libName}/src/lib/${libName}.module.ts`;
  tree.write(
    modulePath,
    `import * as ButtonExports from './test-button/test-button.component';
    ${tree.read(modulePath)}`
  );

  // create a module with component that gets exported in a barrel file
  generateModule(tree, {
    name: 'barrel',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'barrel-button',
    path: `${libName}/src/lib/barrel/barrel-button/barrel-button`,
    module: 'barrel',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/barrel/barrel-button/index.ts`,
    `export * from './barrel-button.component';`
  );

  tree.write(
    `${libName}/src/lib/barrel/barrel.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BarrelButtonComponent } from './barrel-button';

@NgModule({
  imports: [CommonModule],
  declarations: [BarrelButtonComponent],
})
export class BarrelModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  generateModule(tree, {
    name: 'variable-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-button',
    path: `${libName}/src/lib/variable-declare/variable-declare-button/variable-declare-button`,
    module: 'variable-declare',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'variable-declare-view',
    path: `${libName}/src/lib/variable-declare/variable-declare-view/variable-declare-view`,
    module: 'variable-declare',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/variable-declare/variable-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableDeclareButtonComponent } from './variable-declare-button/variable-declare-button.component';
import { VariableDeclareViewComponent } from './variable-declare-view/variable-declare-view.component';

const COMPONENTS = [
  VariableDeclareButtonComponent,
  VariableDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: COMPONENTS,
  exports: COMPONENTS
})
export class VariableDeclareModule {}`
  );

  // create a module with components that get Angular exported and declared by variable
  generateModule(tree, {
    name: 'variable-spread-declare',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-button',
    path: `${libName}/src/lib/variable-spread-declare/variable-spread-declare-button/variable-spread-declare-button`,
    module: 'variable-spread-declare',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-view',
    path: `${libName}/src/lib/variable-spread-declare/variable-spread-declare-view/variable-spread-declare-view`,
    module: 'variable-spread-declare',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'variable-spread-declare-anotherview',
    path: `${libName}/src/lib/variable-spread-declare/variable-spread-declare-anotherview/variable-spread-declare-anotherview`,
    module: 'variable-spread-declare',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/variable-spread-declare/variable-spread-declare.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VariableSpreadDeclareButtonComponent } from './variable-spread-declare-button/variable-spread-declare-button.component';
import { VariableSpreadDeclareViewComponent } from './variable-spread-declare-view/variable-spread-declare-view.component';
import { VariableSpreadDeclareAnotherviewComponent } from './variable-spread-declare-anotherview/variable-spread-declare-anotherview.component';

const COMPONENTS = [
  VariableSpreadDeclareButtonComponent,
  VariableSpreadDeclareViewComponent
]

@NgModule({
  imports: [CommonModule],
  declarations: [...COMPONENTS, VariableSpreadDeclareAnotherviewComponent],
})
export class VariableSpreadDeclareModule {}`
  );

  // create a module where declared components are pulled from a static member of the module
  generateModule(tree, {
    name: 'static-member-declarations',
    project: libName,
  });

  await componentGenerator(tree, {
    name: 'cmp1',
    path: `${libName}/src/lib/static-member-declarations/cmp1/cmp1`,
    module: 'static-member-declarations',
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'cmp2',
    path: `${libName}/src/lib/static-member-declarations/cmp2/cmp2`,
    module: 'static-member-declarations',
    standalone: false,
    skipFormat: true,
  });

  tree.write(
    `${libName}/src/lib/static-member-declarations/static-member-declarations.module.ts`,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Cmp1Component } from './cmp1/cmp1.component';
import { Cmp2Component } from './cmp2/cmp2.component';

@NgModule({
  imports: [CommonModule],
  declarations: StaticMemberDeclarationsModule.COMPONENTS,
  exports: StaticMemberDeclarationsModule.COMPONENTS
})
export class StaticMemberDeclarationsModule {
  static readonly COMPONENTS = [Cmp1Component, Cmp2Component];
}`
  );

  // create another button in a nested subpath
  generateModule(tree, {
    name: 'nested',
    project: libName,
    path: `${libName}/src/lib`,
  });

  await componentGenerator(tree, {
    name: 'nested-button',
    module: 'nested',
    path: `${libName}/src/lib/nested/nested-button/nested-button`,
    standalone: false,
    skipFormat: true,
  });

  await componentGenerator(tree, {
    name: 'test-other',
    path: `${libName}/src/lib/test-other/test-other`,
    standalone: false,
    skipFormat: true,
  });

  return tree;
}

function generateModule(
  tree: Tree,
  options: { name: string; project: string; path?: string }
): void {
  const project = readProjectConfiguration(tree, options.project);

  if (options.path === undefined) {
    const sourceRoot = project.sourceRoot ?? `${project.root}/src`;
    const projectDirName =
      project.projectType === 'application' ? 'app' : 'lib';
    options.path = `${sourceRoot}/${projectDirName}`;
  }

  const moduleNames = names(options.name);
  const moduleFilePath = `${options.path}/${moduleNames.fileName}/${moduleNames.fileName}.module.ts`;

  tree.write(
    moduleFilePath,
    `import { NgModule } from '@angular/core';
  import { CommonModule } from '@angular/common';

  @NgModule({
    declarations: [],
    imports: [CommonModule],
  })
  export class ${moduleNames.className}Module {}`
  );
}
