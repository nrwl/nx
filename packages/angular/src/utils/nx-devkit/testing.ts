import type { Tree } from '@nx/devkit';
import { addProjectConfiguration, names, updateJson } from '@nx/devkit';

export interface AppConfig {
  appName: string; // name of app
  appModule: string; // app/app.module.ts in the above sourceDir
}

export interface LibConfig {
  name: string;
  module: string;
  barrel: string;
}

var appConfig: AppConfig; // configure built in createApp()
var libConfig: LibConfig;

export function getAppConfig(): AppConfig {
  return appConfig;
}

export function getLibConfig(): LibConfig {
  return libConfig;
}

export function createApp(
  tree: Tree,
  appName: string,
  routing: boolean = true
) {
  appName = names(appName).fileName;
  // save for getAppDir() lookup by external *.spec.ts tests
  appConfig = {
    appName,
    appModule: `${appName}/src/app/app.module.ts`,
  };

  tree.write(
    appConfig.appModule,
    `import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
${routing ? "import { RouterModule } from '@angular/router'" : ''};
import { App } from './app';
@NgModule({
  imports: [BrowserModule, ${routing ? 'RouterModule.forRoot([])' : ''}],
  declarations: [App],
  bootstrap: [App]
})
export class AppModule {}
`
  );
  tree.write(
    `${appName}/src/main.ts`,
    `import { platformBrowser } from '@angular/platform-browser';

import { AppModule } from './app/app.module';

platformBrowser()
  .bootstrapModule(AppModule)
  .catch(err => console.log(err));
  `
  );
  tree.write(
    `${appName}/tsconfig.app.json`,
    JSON.stringify({
      include: ['**/*.ts'],
    })
  );
  tree.write(
    `${appName}-e2e/tsconfig.e2e.json`,
    JSON.stringify({
      include: ['../**/*.ts'],
    })
  );
  addProjectConfiguration(tree, appName, {
    root: `${appName}`,
    sourceRoot: `${appName}/src`,
    targets: {
      build: {
        options: {
          main: `${appName}/src/main.ts`,
        },
      },
      serve: {
        options: {},
      },
    },
    tags: [],
  });
}

export function createLib(tree: Tree, libName: string) {
  const { name, className, fileName, propertyName } = names(libName);

  libConfig = {
    name,
    module: `${propertyName}/src/lib/${fileName}.module.ts`,
    barrel: `${propertyName}/src/index.ts`,
  };

  tree.write(
    libConfig.module,
    `import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
@NgModule({
  imports: [
    CommonModule
  ],
  providers: []
})
export class ${className}Module { }
  `
  );
  tree.write(
    libConfig.barrel,
    `export * from './lib/${fileName}.module';
`
  );
}
