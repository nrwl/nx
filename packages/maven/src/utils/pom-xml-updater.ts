import { Tree } from '@nx/devkit';
import { DOMParser, XMLSerializer } from '@xmldom/xmldom';

/**
 * Updates the version of dev.nx.maven:nx-maven-plugin in pom.xml files.
 * Only updates the version element that belongs to the nx-maven-plugin.
 *
 * @param tree - Nx Tree instance
 * @param version - The new version to set
 */
export function updateNxMavenPluginVersion(tree: Tree, version: string): void {
  const pomPath = 'pom.xml';

  if (!tree.exists(pomPath)) {
    return;
  }

  const content = tree.read(pomPath, 'utf-8');
  const parser = new DOMParser();
  const doc = parser.parseFromString(content);

  let hasChanges = false;

  // Find all plugin elements in the document
  const plugins = Array.from(doc.getElementsByTagName('plugin'));

  for (const plugin of plugins) {
    // Check if this is the nx-maven-plugin by looking for groupId and artifactId
    const groupId = plugin.getElementsByTagName('groupId')[0];
    const artifactId = plugin.getElementsByTagName('artifactId')[0];

    const isNxMavenPlugin =
      groupId?.textContent?.trim() === 'dev.nx.maven' &&
      artifactId?.textContent?.trim() === 'nx-maven-plugin';

    // If this is the nx-maven-plugin, update its version
    if (isNxMavenPlugin) {
      const versionElement = plugin.getElementsByTagName('version')[0];
      const currentVersion = versionElement?.textContent?.trim();

      if (currentVersion && currentVersion !== version) {
        versionElement.textContent = version;
        hasChanges = true;
      }
    }
  }

  // If content changed, write it back using XML serializer to preserve formatting
  if (hasChanges) {
    const serializer = new XMLSerializer();
    const updatedContent = serializer.serializeToString(doc);
    tree.write(pomPath, updatedContent);
  }
}
