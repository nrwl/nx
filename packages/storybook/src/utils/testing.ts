import { Tree } from '@nx/devkit';

import { libraryGenerator } from '@nx/js';
import { Linter } from '@nx/eslint';

export async function createTestUILibNoNgDevkit(
  appTree: Tree,
  libName: string
): Promise<Tree> {
  await libraryGenerator(appTree, {
    linter: Linter.EsLint,
    skipFormat: true,
    skipTsConfig: false,
    unitTestRunner: 'none',
    directory: libName,
  });

  return appTree;
}

export function deleteNewConfigurationAndCreateNew(
  appTree: Tree,
  projectStorybookRoot: string
): Tree {
  // Remove new Storybook configuration
  appTree.delete(`.storybook/main.js`);
  appTree.delete(`.storybook/tsconfig.json`);
  appTree.delete(`${projectStorybookRoot}/main.js`);
  appTree.delete(`${projectStorybookRoot}/preview.js`);
  appTree.delete(`${projectStorybookRoot}/tsconfig.json`);

  // Create old Storybook configuration
  appTree.write(`.storybook/addons.js`, 'console.log("hello")');
  appTree.write(`.storybook/webpack.config.js`, 'console.log("hello")');
  appTree.write(`.storybook/tsconfig.json`, '{"test": "hello"}');
  appTree.write(`${projectStorybookRoot}/config.js`, 'console.log("hello")');
  appTree.write(`${projectStorybookRoot}/addons.js`, 'console.log("hello")');
  appTree.write(
    `${projectStorybookRoot}/webpack.config.js`,
    'console.log("hello")'
  );
  appTree.write(`${projectStorybookRoot}/tsconfig.json`, '{"test": "hello"}');

  return appTree;
}
