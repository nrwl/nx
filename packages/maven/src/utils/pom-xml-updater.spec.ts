import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { updateNxMavenPluginVersion } from './pom-xml-updater';

describe('pom-xml-updater', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({});
  });

  describe('updateNxMavenPluginVersion', () => {
    it('should update the nx-maven-plugin version only', () => {
      // Arrange
      const pomContent = `<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.7</version>
      </plugin>
    </plugins>
  </build>
</project>`;

      tree.write('pom.xml', pomContent);

      // Act
      updateNxMavenPluginVersion(tree, '0.0.8');

      // Assert
      const updatedContent = tree.read('pom.xml', 'utf-8');
      expect(updatedContent).toMatchSnapshot();
    });

    it('should not update other plugin versions', () => {
      // Arrange
      const pomContent = `<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.7</version>
      </plugin>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>0.0.7</version>
      </plugin>
    </plugins>
  </build>
</project>`;

      tree.write('pom.xml', pomContent);

      // Act
      updateNxMavenPluginVersion(tree, '0.0.8');

      // Assert
      const updatedContent = tree.read('pom.xml', 'utf-8');
      expect(updatedContent).toMatchSnapshot();
    });

    it('should not modify pom.xml if file does not exist', () => {
      // Act & Assert - should not throw
      expect(() => {
        updateNxMavenPluginVersion(tree, '0.0.8');
      }).not.toThrow();
    });

    it('should preserve XML formatting and structure', () => {
      // Arrange
      const pomContent = `<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.7</version>
      </plugin>
    </plugins>
  </build>
</project>`;

      tree.write('pom.xml', pomContent);

      // Act
      updateNxMavenPluginVersion(tree, '0.0.8');

      // Assert
      const updatedContent = tree.read('pom.xml', 'utf-8');
      expect(updatedContent).toMatchSnapshot();
    });

    it('should not update if version is already the target version', () => {
      // Arrange
      const pomContent = `<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.8</version>
      </plugin>
    </plugins>
  </build>
</project>`;

      tree.write('pom.xml', pomContent);

      // Act
      updateNxMavenPluginVersion(tree, '0.0.8');

      // Assert
      const updatedContent = tree.read('pom.xml', 'utf-8');
      expect(updatedContent).toEqual(pomContent);
    });

    it('should handle pom.xml with dependencies and other version elements', () => {
      // Arrange
      const pomContent = `<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <version>1.0.0</version>
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.7</version>
      </plugin>
    </plugins>
  </build>
  <dependencies>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.13.2</version>
    </dependency>
  </dependencies>
</project>`;

      tree.write('pom.xml', pomContent);

      // Act
      updateNxMavenPluginVersion(tree, '0.0.8');

      // Assert
      const updatedContent = tree.read('pom.xml', 'utf-8');
      expect(updatedContent).toMatchSnapshot();
    });

    it('should correctly identify nx-maven-plugin with whitespace', () => {
      // Arrange
      const pomContent = `<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <build>
    <plugins>
      <plugin>
        <groupId>
          dev.nx.maven
        </groupId>
        <artifactId>
          nx-maven-plugin
        </artifactId>
        <version>
          0.0.7
        </version>
      </plugin>
    </plugins>
  </build>
</project>`;

      tree.write('pom.xml', pomContent);

      // Act
      updateNxMavenPluginVersion(tree, '0.0.8');

      // Assert
      const updatedContent = tree.read('pom.xml', 'utf-8');
      expect(updatedContent).toMatchSnapshot();
    });

    it('should only update plugins, not parent version', () => {
      // Arrange
      const pomContent = `<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <parent>
    <groupId>dev.nx</groupId>
    <artifactId>nx-parent</artifactId>
    <version>0.0.7</version>
  </parent>
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.7</version>
      </plugin>
    </plugins>
  </build>
</project>`;

      tree.write('pom.xml', pomContent);

      // Act
      updateNxMavenPluginVersion(tree, '0.0.8');

      // Assert
      const updatedContent = tree.read('pom.xml', 'utf-8');
      expect(updatedContent).toMatchSnapshot();
    });
  });
});
