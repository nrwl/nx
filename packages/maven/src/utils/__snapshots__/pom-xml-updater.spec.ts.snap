// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`pom-xml-updater updateNxMavenPluginVersion should correctly identify nx-maven-plugin with whitespace 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <build>
    <plugins>
      <plugin>
        <groupId>
          dev.nx.maven
        </groupId>
        <artifactId>
          nx-maven-plugin
        </artifactId>
        <version>0.0.8</version>
      </plugin>
    </plugins>
  </build>
</project>"
`;

exports[`pom-xml-updater updateNxMavenPluginVersion should handle pom.xml with dependencies and other version elements 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <version>1.0.0</version>
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.8</version>
      </plugin>
    </plugins>
  </build>
  <dependencies>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.13.2</version>
    </dependency>
  </dependencies>
</project>"
`;

exports[`pom-xml-updater updateNxMavenPluginVersion should not update other plugin versions 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.8</version>
      </plugin>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>0.0.7</version>
      </plugin>
    </plugins>
  </build>
</project>"
`;

exports[`pom-xml-updater updateNxMavenPluginVersion should only update plugins, not parent version 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <parent>
    <groupId>dev.nx</groupId>
    <artifactId>nx-parent</artifactId>
    <version>0.0.7</version>
  </parent>
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.8</version>
      </plugin>
    </plugins>
  </build>
</project>"
`;

exports[`pom-xml-updater updateNxMavenPluginVersion should preserve XML formatting and structure 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.8</version>
      </plugin>
    </plugins>
  </build>
</project>"
`;

exports[`pom-xml-updater updateNxMavenPluginVersion should update the nx-maven-plugin version only 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0">
  <build>
    <plugins>
      <plugin>
        <groupId>dev.nx.maven</groupId>
        <artifactId>nx-maven-plugin</artifactId>
        <version>0.0.8</version>
      </plugin>
    </plugins>
  </build>
</project>"
`;
